### PR TITLE
Review feedback and several cleanups

### PR DESCRIPTION
**Pull request by s1monw**
@javanna @martijnvg FYI

### CODE REVIEW

**Comment by s1monw** [comment]
I will do another review once we got rid of the index as a storage - looks pretty good already guys!

File: core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java [hunk 1]
```diff
@@ -16,7 +16,7 @@
  * specific language governing permissions and limitations
  * under the License.
  */
-
+//TODO(simonw): can all these classes go into org.elasticsearch.ingest?
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java]: the idea would be to keep two packages as the .core package should be es agnostic, while the ingest one is the actual runtime/infra that heavily depends on core. I would like to keep this distinction one way or another, maybe there's other way to enforce that we don't bloat with es dependencies these few classes like IngestDocument, Pipeline and the processors. Ideas?

File: core/src/main/java/org/elasticsearch/action/ingest/SimulateExecutionService.java [hunk 2]
```diff
@@ -65,12 +66,15 @@ SimulateDocumentResult executeDocument(Pipeline pipeline, IngestDocument ingestD
...
-            for (IngestDocument ingestDocument : request.getDocuments()) {
-                responses.add(executeDocument(request.getPipeline(), ingestDocument, request.isVerbose()));
+        threadPool.executor(THREAD_POOL_NAME).execute(new ActionRunnable<SimulatePipelineResponse>(listener) {
+            @Override
+            protected void doRun() throws Exception {
+                List<SimulateDocumentResult> responses = new ArrayList<>();
+                for (IngestDocument ingestDocument : request.getDocuments()) {
+                    responses.add(executeDocument(request.getPipeline(), ingestDocument, request.isVerbose()));
+                }
+                listener.onResponse(new SimulatePipelineResponse(request.getPipeline().getId(), request.isVerbose(), responses));
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/ingest/SimulateExecutionService.java]: for my education, what is the reason of this change? what does it buy us?

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/ingest/SimulateExecutionService.java]: if you just use a runnable and the execution fails or the task gets rejected we never call the listener?

File: core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java]: why can't we just have one package I don't get it?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java]: we can... but it's important to not forget that some classes should not to depend on es. Packages don't enforce anything but may make you think about it (or maybe not). Anyways I think it's nice to keep separated the es runtime for ingest and the ingest classes that are independent from es core. There is a pretty big difference between these two sets of classes. As I asked above, maybe there are other ways to keep things clean, not sure.

File: core/src/main/java/org/elasticsearch/action/ingest/SimulateExecutionService.java (hunk 2 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/ingest/SimulateExecutionService.java]: gotcha

File: core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java]: the only way I see is to make ES depend on it an make it a sep jar just luke test-framework

Comment (javanna) [inline core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java]: agreed, I guess we are not ready for that yet but we will do it at some point? If that is the plan we should still keep separated the classes that will eventually go out? This was the only purpose of the separate package.

Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java]: Maybe for now we should move things into one package. When we want to make ingest a separate jar then this will be a separate project. There are other things that we will have to address  once we move to a dedicated ingest jar.

**Comment by martijnvg** [comment]
+1 for the cleanups. We will address the TODO's in followup PRs.

File: core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java (hunk 1 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ingest/core/CompoundProcessor.java]: I would be more in favour of keeping the two packages for now :) logic separation seems likes a good enough reason to keep two packages, that doesn't mean that the core package can already go out with no changes but the changes required should be small-ish, while I am afraid if we merge everything into a single package we will quickly lose this distinction (think of making IngestDocument and Pipeline writeable instead of having separate classes for that), that feels like undoing some of the work we have already done.

**Comment by s1monw** [comment]
@martijnvg should i just merge it then?

**Comment by javanna** [comment]
yes @s1monw merge please, we will take it from here.

**Comment by martijnvg** [comment]
@s1monw yes, +1 to merge
